### PR TITLE
feat/EN-3483-merge-header-interceptors

### PR DESCRIPTION
**Pull request by camilbancioiu**
The following structs can be merged into a single struct, since they do the same thing:

* `HeaderInterceptor`
* `HeaderInterceptorBase`
* `ShardHeaderInterceptor`

This pull request does the following:
1. Merge the code from `HeaderInterceptorBase` into `HeaderInterceptor`, then delete `HeaderInterceptorBase`
2. Delete `ShardHeaderInterceptor`
3. Replace `metablock.NewShardHeaderInterceptor` with `interceptors.NewHeaderInterceptor` in the only place it was called.

### CODE REVIEW

File: process/block/interceptors/headerInterceptor.go [hunk 1]
```diff
@@ -34,38 +38,81 @@ func NewHeaderInterceptor(
...
+		headers:             headers,
+		headersNonces:       headersNonces,
+	}
+
+	return hdrInterceptor, nil
+}
+
+// ParseReceivedMessage will transform the received p2p.Message in an InterceptedHeader.
+// If the header hash is present in storage it will output an error
+func (hi *HeaderInterceptor) ParseReceivedMessage(message p2p.MessageP2P) (*block.InterceptedHeader, error) {
```
Comment (sasurobert) [inline process/block/interceptors/headerInterceptor.go]: improve code coverage, copy test functions from shard and base test to headerInterceptor_test to test every bit.

Comment (camilbancioiu) [reply process/block/interceptors/headerInterceptor.go]: Damn. Forgot that the tests needed merging as well.

File: process/block/interceptors/headerInterceptor.go [hunk 2]
```diff
@@ -75,12 +122,20 @@ func (hi *HeaderInterceptor) ProcessReceivedMessage(message p2p.MessageP2P) erro
 	return nil
 }
 
+// CheckHeaderForCurrentShard checks if the header is for current shard
+func (hi *HeaderInterceptor) CheckHeaderForCurrentShard(interceptedHdr *block.InterceptedHeader) bool {
```
Comment (iulianpascalau) [inline process/block/interceptors/headerInterceptor.go]: Maybe make this unexported?

File: process/block/interceptors/headerInterceptor.go (hunk 1 above)
Comment (camilbancioiu) [reply process/block/interceptors/headerInterceptor.go]: Did the merging. Now `headerInterceptor_test.go` contains the union of all the tests for these 3 structs, but without any duplicates. I'm not entirely sure that I unified the tests correctly - I don't understand the `headerInterceptor` 100%.

File: process/block/interceptors/headerInterceptor.go (hunk 2 above)
Comment (camilbancioiu) [reply process/block/interceptors/headerInterceptor.go]: Changed it.

File: process/block/interceptors/headerInterceptor.go (hunk 1 above)
Comment (iulianpascalau) [reply process/block/interceptors/headerInterceptor.go]: Tests seems fine, should add more tests for the contructor

File: process/block/interceptors/headerInterceptor.go (hunk 2 above)
Comment (iulianpascalau) [reply process/block/interceptors/headerInterceptor.go]: Ok, please update the comment also

File: process/block/interceptors/headerInterceptor.go (hunk 1 above)
Comment (camilbancioiu) [reply process/block/interceptors/headerInterceptor.go]: Added extra tests for the constructor.

File: process/block/interceptors/headerInterceptor.go (hunk 2 above)
Comment (camilbancioiu) [reply process/block/interceptors/headerInterceptor.go]: Done.

**Pull request merged**